cEP-0033.md

```diff
@@ -0,0 +1,760 @@
+                             than start as negative ranges are not allowed.
+        :param language:     The programming language of the lines.
+        :param index:        The index of the nl_section.
+        :param index:        The index of the nl_section.
+        :param linted_start: The start line number in the linted file.
+        :param linted_end:   The end line number in the linted file.
+        :param temp_file:    The name of the temporary segragated file to which 
+                             the section is associated with.
+
+                            
+        :raises ValueError: Raised when file of start and end mismatch.
+        """
+        self.index = index
+        self.language = language
+        self.orig_start = start
+        self.orig_end = end
+        self.linted_start = linted_start
+        self.linted_end = linted_end
+
+        if self.start.file != self.end.file:
+            raise ValueError('File of start and end position do not match.')
+
+    @classmethod
+    def from_values(cls,
+                    start_line=None,
+                    end_line=None,
+                    language=None,
+                    index=None):
+        start = NlSection(start_line, end_line, language, index)
+        if end_line:
+            end = NlSection(start_line, end_line if end_line else start_line)
+        else:
+            end = None
+
+        return cls(start, end, language, index)
+
+```
+
+### Note
+
+The implementation of the nested language architecture would be done inside the
+a new folder called as `nestedlib` . This `nestedlib` would be present under
+the `coalib` directory.
+
+
+## Architecture
+
+The following figure depicts the architecture that is going to be implemented
+to enable nested languages support.
+
+![Architecture](https://github.com/Naveenaidu/gsoc-18-materials/blob/master/photos/arch-detailed.jpg)
+
+* __NLCore__
+
+This is the hear of the project. It is responsible to manage the execution flow
+in the program
+
+* __NLInfoExtractor__
+
+NLInfoExtractor is responsible to extract the information from the user's input. 
+The extracted information would include the languages present in the file,
+bears to run for each programming language and the settings of these bears.
+
+* __Parser__
+
+Parser splits up the original nested file into different `nl_sections`.
+
+* __NLFileHandler__
+
+NL FileHandler is responsible for creating the temporary files (_which contains
+all the sections belonging to a single language_) that would be passed for the 
+actual analysis.
+
+## Implementation
+
+This project will be implemented in four phases:
+
+* Information Gathering
+* Segregating the file
+* Linting the file
+* Assembling
+
+## Phase 1: Information Gathering
+
+The users can inform coala about the presence of nested languages in the file
+using the `--handle-nested` flag.
+
+eg:
+```bash
+coala --handle-nested --file=html_php_file.html --language=html,php --bears=HTMLBear,PHPBear 
```

Comment:
not `--files`? (note the plural-'s')

Reply:
A typo, we can put `--files` too, the architecture will support any number of files provided that the both the  files have the same nested languages.

Will be updating it.